### ISSUE TITLE
Infinite loop when Suspense contains a Provider with initialValues set

### ISSUE THREAD

**Issue by macarie** [pmndrs/jotai#1112]
I played around with Jotai to create a (custom) translation library.

We need to show some app components using a different locale, so I thought we could do this by defining a `scope` as `useAtom`'s second argument and providing the locale through `Provider`'s `initialValues` and `scope`.

I have two `atom`s: `localeAtom` and `translationsAtom`.

`translationsAtom` is async and does a `get(localeAtom)`, then checks if the translations for that `locale` are already cached. If they are, it returns them; otherwise, it imports them, sets the cache, and returns them.

Everything works fine. I can change the `localeAtom`'s value, and the `translationsAtom`'s value changes.

The problem arises when I use the `Provider` component and set `initialValues`; at this point, `translationsAtom` enters an infinite loop.

I've set up a [barebones example here](https://codesandbox.io/s/jotai-loop-async-atom-provider-pcnnh1?file=/src/index.tsx), without the custom hook implementation and the use of `scope` for simplicity's sake. I've left some logs in there to see where it falls apart. To reproduce the issue is enough to uncomment the lines where the `Provider` is declared (`index.tsx:13` and `index.tsx:15` if the link doesn't open the correct file).

**Comment by dai-shi** [comment]
The example is too long, but I guess the problem is `<Suspense>` is not inside `<Provider>`. Hope it helps.

**Comment by macarie** [comment]
Oh, yeah, that was the problem!

I've just found the paragraph about that in the docs. I think I've missed it before.

Thanks for the help, and sorry for having wasted your time 😅

Do you think it would be possible to detect this scenario and add a warn/error to the console with a link to the docs, like React does for some common issues? Having an infinite loop doesn't feel right, and honestly, moving the `Suspense` component inside the `Provider` didn't even cross my mind, but this might be because `Suspense` is kinda new.

**Comment by dai-shi** [comment]
Infinite loop detection would generally be hard. Maybe possible with some heuristics. If anyone has a not-too-complex idea, please let us know.

What we should do at least is to improve the docs to explicitly tell (or even warn) the usage of `Suspense`. Do you have any suggestion?

**Comment by macarie** [comment]
I've tried debugging this issue because I thought it was weird that it doesn't work only with `initialValues`.

I've found out that what causes the infinite loop is this line of code: https://github.com/pmndrs/jotai/blob/3f7d555355bba3d8743915a8b7f8fda4e0eb0991/src/core/Provider.ts#L46

I think, but I might be wrong on this, that in my case (`Suspense > Provider > useAtom(async)`), the `Provider` is mounted/created multiple times.

Whenever this happens, the code creates a new value for `scopeContainerRef` that's used as `ScopeContainerContext`'s value, thus creating the infinite loop. `ScopeContainerContext` is already cached using `scope` as a key so that it won't make a new `Context` each time; it just keeps changing its value every time, thus re-rendering the component because of the `useContext` used here: https://github.com/pmndrs/jotai/blob/3f7d555355bba3d8743915a8b7f8fda4e0eb0991/src/core/useAtomValue.ts#L21

I've tried to use [object-hash](https://www.npmjs.com/package/object-hash) to create stable hashes (based on key-value pairs instead of reference) for `initialValues`, caching `createScopeContainer`'s result, and using that as `ScopeContainerContext`'s value. This seems to fix everything.

<details>
<summary>These are my (<em>not so well thought</em>*) changes:</summary>

```diff
diff --git a/src/core/Provider.ts b/src/core/Provider.ts
index c806e9d..40423bb 100644
--- a/src/core/Provider.ts
+++ b/src/core/Provider.ts
@@ -1,3 +1,4 @@
+import hash from 'object-hash'
 import { createElement, useEffect, useRef, useState } from 'react'
 import type { PropsWithChildren } from 'react'
 import type { Atom, Scope } from './atom'
@@ -7,6 +8,28 @@ import { COMMIT_ATOM, createStoreForExport } from './store'
 import type { VersionObject } from './store'
 import { useDebugState } from './useDebugState'
 
+const cachedScopeContainers = new Map<string | undefined, ScopeContainer>()
+const retrieveOrCreateScopeContainer = (
+  initialValues: Iterable<readonly [Atom<unknown>, unknown]> | undefined,
+  unstable_createStore: typeof createStoreForExport | undefined
+): ScopeContainer => {
+  const cacheKey: string | undefined =
+    initialValues !== undefined ? hash(initialValues) : initialValues
+
+  if (cachedScopeContainers.has(cacheKey)) {
+    return cachedScopeContainers.get(cacheKey)!
+  }
+
+  const scopeContainer = createScopeContainer(
+    initialValues,
+    unstable_createStore
+  )
+
+  cachedScopeContainers.set(cacheKey, scopeContainer)
+
+  return scopeContainer
+}
+
 export const Provider = ({
   children,
   initialValues,
@@ -43,10 +66,11 @@ export const Provider = ({
   const scopeContainerRef = useRef<ScopeContainer>()
   if (!scopeContainerRef.current) {
     // lazy initialization
-    scopeContainerRef.current = createScopeContainer(
+    scopeContainerRef.current = retrieveOrCreateScopeContainer(
       initialValues,
       unstable_createStore
     )
+
     if (unstable_enableVersionedWrite) {
       scopeContainerRef.current.w = (write) => {
         setVersion((parentVersion) => {
``` 
</details>

<details>
<summary>And here's a my use case as a test:</summary>

```diff
diff --git a/tests/async.test.tsx b/tests/async.test.tsx
index 195aef4..acb1cf3 100644
--- a/tests/async.test.tsx
+++ b/tests/async.test.tsx
@@ -1137,3 +1137,37 @@ it('multiple derived atoms with dependency chaining and async write (#813)', asy
     getByText('bName: beta')
   })
 })
+
+it('Provider with initialValues inside Suspense should not cause infinite loop', async () => {
+  const wait = (t: number): Promise<void> =>
+    new Promise((resolve) => setTimeout(resolve, t))
+  const countAtom = atom(0)
+  const asyncAtom = atom(async (get) => {
+    const counter = get(countAtom)
+
+    await wait(counter)
+
+    return counter
+  })
+
+  const scope = 'atomScope'
+  const intialCount = 42
+
+  const Display = () => {
+    const [r] = useAtom(asyncAtom, scope)
+
+    return <p>count: {r}</p>
+  }
+
+  const { getByText } = render(
+    <Suspense fallback="loading">
+      <Provider scope={scope} initialValues={[[countAtom, intialCount]]}>
+        <Display />
+      </Provider>
+    </Suspense>
+  )
+
+  await waitFor(() => {
+    getByText(`count: ${intialCount}`)
+  })
+}, 100)
```
</details>

At this point, I don't think it's a good idea to support this use case, given the complexity. A warning on the console or interrupting the loop would be nice, but it would require creating a hash of `initialValues` anyway.

As a counter-argument, supporting this would allow having a single `Suspense` for multiple atom consumers, each having its own `Provider`, instead of using various `Suspense`s.

> What we should do at least is to improve the docs to explicitly tell (or even warn) the usage of Suspense. Do you have any suggestion?

The documentation looks good; I would highlight the quote below more. Maybe it should be a paragraph with a title?

> If you have `<Provider>`, place at least one `<Suspense>` inside the `<Provider>`.

Again, sorry for the long post, I'm not 100% sure this is the root of the problem, but it looks like it.

---

*: they break this test: https://github.com/pmndrs/jotai/blob/3f7d555355bba3d8743915a8b7f8fda4e0eb0991/tests/utils/useHydrateAtoms.test.tsx#L216

**Comment by dai-shi** [comment]
The current behavior (re-creating a store on remount) is not a bug or something to fix, but it's an expected/intentional behavior. Otherwise, it will break many use cases. It's how React state works in general.

> The documentation looks good; I would highlight the quote below more. Maybe it should be a paragraph with a title?

Would you please send a PR? That will be helpful!

----

A side note:

> As a counter-argument, supporting this would allow having a single Suspense for multiple atom consumers, each having its own Provider, instead of using various Suspenses.

It is not directly related to your point, but [jotai-molecules](https://github.com/saasquatch/jotai-molecules) might be interesting with your mental model.

**Comment by macarie** [comment]
> The current behavior (re-creating a store on remount) is not a bug or something to fix, but it's an expected/intentional behavior. Otherwise, it will break many use cases. It's how React state works in general.

Sorry, I think I haven't worded it correctly. I agree with you; this (for me) is more like a weird and complex edge case that Jotai should not support 👍🏼

I had some free time and wanted to write down what I've found so that if other people have the same problem, they can better understand the cause.

> Would you please send a PR? That will be helpful!

Sure!

> It is not directly related to your point, but [jotai-molecules](https://github.com/saasquatch/jotai-molecules) might be interesting with your mental model.

Thanks, will take a look 😄

**Comment by dai-shi** [comment]
closed by #1142.
